crates/cheatnet/src/cheatcodes/deploy.rs

```diff
@@ -25,15 +25,39 @@ impl CheatnetState {
         &mut self,
         class_hash: &ClassHash,
         calldata: &[Felt252],
+        predefined_contract_address: Option<ContractAddress>,
     ) -> Result<ContractAddress, CheatcodeError> {
         // Deploy a contract using syscall deploy.
         let account_address = ContractAddress(patricia_key!(TEST_ACCOUNT_CONTRACT_ADDRESS));
         let entry_point_selector = selector_from_name("deploy_contract");
         let salt = self.get_salt();
-        let contract_address = self.precalculate_address(class_hash, calldata);
+
+        let contract_address = if let Some(contract_address) = predefined_contract_address {
+            if self
+                .blockifier_state
+                .state
+                .address_to_class_hash
+                .get(&contract_address)
+                .is_some()
+            {
+                return Err(CheatcodeError::Recoverable(vec![felt_from_short_string(
+                    "Address is already taken",
+                )]));
+            }
+
+            contract_address
+        } else {
+            self.precalculate_address(class_hash, calldata)
+        };
         self.increment_deploy_salt_base();
 
         let blockifier_state: &mut CachedState<DictStateReader> = &mut self.blockifier_state;
+        if predefined_contract_address.is_some() {
```

Comment:
Do we need the `if`? Cannot we always insert that as we already assigned to `contract_address`?

Reply:
Unfortunately not. If we do so there will be an error in `call_contract` that the address is already taken